### PR TITLE
New controller board 'kintwin' for Kinesis Contoured keyboards

### PR DESCRIPTION
**Pull request by alvicstep**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kinesis/kintwin/info.json
Comment (dunk2k) [inline keyboards/kinesis/kintwin/info.json:13]: migrated from `rules.mk`
```suggestion
    "usb": {
        "device_version": "0.0.3",
        "vid": "0xFEED",
        "pid": "0x4B4E",
        "shared_endpoint": {
            "keyboard": true
        }
    },
    "features": {
        "bootmagic": true,
        "console": false,
        "command": false,
        "mousekey": false,
        "extrakey": false,
        "nkro": true,
    },
```

File: keyboards/kinesis/kintwin/readme.md
Comment (dunk2k) [inline keyboards/kinesis/kintwin/readme.md:13]:
```suggestion
* Hardware Supported: KinTwin PCB, WeAct Black Pill STM32F401 micro controller
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/kinesis/kintwin/info.json
Comment (dunk2k) [inline keyboards/kinesis/kintwin/info.json:22]: looking at similar kinesis conversions, is `num_lock` missing here?

Comment (alvicstep) [reply keyboards/kinesis/kintwin/info.json:31]: will add both num_lock and compose

File: keyboards/kinesis/kintwin/rules.mk
Comment (alvicstep) [reply keyboards/kinesis/kintwin/rules.mk]: I will revert this change for now.

Comment (dunk2k) [reply keyboards/kinesis/kintwin/rules.mk]: Would the build warnings be indicative that each controller/board under `kinesis` dir should have their own rules.mk/info.json defining what feature are enabled/disabled, rather than having a blanket/master rules.mk?

Comment (dunk2k) [reply keyboards/kinesis/kintwin/rules.mk]: Linting check is failing due to no licence headers in identified keymaps.

Comment (alvicstep) [reply keyboards/kinesis/kintwin/rules.mk]: > Would the build warnings be indicative that each controller/board under `kinesis` dir should have their own rules.mk/info.json defining what feature are enabled/disabled, rather than having a blanket/master rules.mk?

Yes, I think master rules.mk is unnecessary. But its removal might result in build option changes for some controllers.

Comment (alvicstep) [reply keyboards/kinesis/kintwin/rules.mk]: > 

What files are missing license headers?

Comment (dunk2k) [reply keyboards/kinesis/kintwin/rules.mk]: > > 
> 
> What files are missing license headers?

`kinesis/keymaps/default/keymap.c` and `kinesis/keymaps/default_pretty/keymap.c`

Comment (dunk2k) [reply keyboards/kinesis/kintwin/rules.mk]: > Yes, I think master rules.mk is unnecessary. But its removal might result in build option changes for some controllers.

Having had a look at `kinesis/rules mk`, thankfully it doesn't contain definitions nor dependencies that would break controllers.

Comment (alvicstep) [reply keyboards/kinesis/kintwin/rules.mk]: Committed requested changes

File: keyboards/kinesis/keymaps/soft_dev/readme.md
Comment (drashna) [inline keyboards/kinesis/keymaps/soft_dev/readme.md:30]:
```suggestion
![](https://i.imgur.com/86LCI3zh.jpg)
![](https://i.imgur.com/Gv4WBw7h.jpg) 
![](https://i.imgur.com/gKXMUI2h.jpg) 
```

File: keyboards/kinesis/kintwin/rules.mk
Comment (dunk2k) [reply keyboards/kinesis/kintwin/rules.mk]: > Committed requested changes

excellent! I'm not sure why change in the most recent commit aven't been reflected in the CI Build, though recently Pull Requests in GitHub (not just qmk_firmware repo) haven't been functioning correctly.
By chance, have you built this firmware locally/on your computer to see if build errors/warning occur?

File: keyboards/kinesis/keymaps/default/keymap.c
Comment (dunk2k) [inline keyboards/kinesis/keymaps/default/keymap.c:13]: whitespace is causing lint check(s) to fail
```suggestion
/* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
```

File: keyboards/kinesis/keymaps/default_pretty/keymap.c
Comment (dunk2k) [inline keyboards/kinesis/keymaps/default_pretty/keymap.c:13]: whitespace causing lint checks to fail
```suggestion
/* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/kinesis/keymaps/default_pretty/keymap.c
Comment (alvicstep) [reply keyboards/kinesis/keymaps/default_pretty/keymap.c]: fixed

File: keyboards/kinesis/keymaps/default/keymap.c
Comment (alvicstep) [reply keyboards/kinesis/keymaps/default/keymap.c]: fixed

**Review by dunk2k — APPROVED** [review]

**Comment by alvicstep** [comment]
I see build errors. It seems these issues were already in the code base even before my PR. 

I see a lot of issues from  ericgebhart dir, like this one:

users/ericgebhart/extensions/keycodes.h:35:47: error: unknown type name 'keyrecord_t'; did you mean 'keyevent_t'?
 bool process_record_secrets(uint16_t keycode, keyrecord_t *record);

How do you usually go about fixing them? Do you try reach out to the maintainer or this is  a responsibility of the last person commiting the code?

**Comment by dunk2k** [comment]
Those are either keymaps or userspace keymaps that have bad code in them. For adding a new keybaord, the important keymap that is require to pass CI build chrcks is `default` (and any "default" community layouts added).

You're not required to fix keymaps or userspace keymaps that aren't yours 🙂

The CI build check helps reviewers determine bad PRs, so a PR with a failed CI build isn't automatically disqualified/ignored.

**Comment by alvicstep** [comment]
Great, thanks

**Review by drashna — APPROVED** [review]

**Comment by laffed** [comment]
@dunk2k @alvicstep 
any update on this MR?
workflow is failing on unrelated/existing problems in other keymaps correct?

**Comment by dunk2k** [comment]
@laffed 
My previous comment, starting "Those are either", still applies.

**Comment by laffed** [comment]
@dunk2k so to recap, 
1. due to some kinesis directory changes, some of the keymaps are now failing build, but not any defaults (that i see)
2. the failing keymaps are not part of the scope of this MR (other maintainers)

So in order for this MR to go through, either @alvicstep would need to fix the failures in this MR or wait until other maintainers fix problems with their own builds. 

But these builds are only failing now in this MR, so these maintainers would not even know this is an issue

**Comment by dunk2k** [comment]
> The CI build check helps reviewers determine bad PRs, so a PR with a failed CI build isn't automatically disqualified/ignored.

**Comment by laffed** [comment]
> @laffed
> 
> > The CI build check helps reviewers determine bad PRs, so a PR with a failed CI build isn't automatically disqualified/ignored.

forgive me, im pretty ignorant here. just a keyboard user who is using the kintwin 

to be crystal clear are you saying:
A. This MR is failing builds, but can be merged in anyways?
B. This MR still needs work by @alvicstep directly related to the features he/she introduced?
C. This MR is waiting for upstream changes by other maintainers before it can be merged?

Who's court is this MR ball currently in?

**Comment by dunk2k** [comment]
> forgive me, im pretty ignorant here. just a keyboard user who is using the kintwin
> 
> to be crystal clear are you saying: A. This MR is failing builds, but can be merged in anyways? B. This MR still needs work by @alvicstep directly related to the features he/she introduced? C. This MR is waiting for upstream changes by other maintainers before it can be merged?
> 
> Who's court is this MR ball currently in?
A. Yes a Pull Request that fails CI Build can still be merged.  
B. There aren't any pending change requests so no further changes/commits are required.
C. This PR has been approved by 1 collaborater and has been passed onto the collaborator team for their review. 

Current status of this PR is it's ready for review and to be merged by collaborator(s), i.e. ball's in collaborator's court.

**Comment by laffed** [comment]
@fauxpark bumping kindly

File: keyboards/kinesis/keymaps/soft_dev/keymap.c
Comment (drashna) [inline keyboards/kinesis/keymaps/soft_dev/keymap.c:49]: This is a nitpick, so feel free to ignore, but this bothers me.

Specifically, most/all keycodes have a 7 character name/alias, so you can have a 9 wide column (eg `1234567, `), so you can have consistent formatting/placement.  So this scrunched together could be better spaced.

**Review by drashna — APPROVED** [review]

File: keyboards/kinesis/kintwin/readme.md
Comment (fauxpark) [inline keyboards/kinesis/kintwin/readme.md:22]:
```suggestion
    make kinesis/kintwin:default

Flashing example for this keyboard:

    make kinesis/kintwin:default:flash
```

File: keyboards/kinesis/kintwin/config.h
Comment (fauxpark) [inline keyboards/kinesis/kintwin/config.h:24]: These can go do info.json.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/kinesis/kintwin/config.h
Comment (alvicstep) [reply keyboards/kinesis/kintwin/config.h]: done

File: keyboards/kinesis/kintwin/info.json
Comment (fauxpark) [inline keyboards/kinesis/kintwin/info.json:36]: Please remove this layout and replace it with `LAYOUT_pretty`.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/kinesis/kintwin/info.json
Comment (alvicstep) [reply keyboards/kinesis/kintwin/info.json:36]: "Layout" matrix was added to have the widest compatibility with the existing keymaps. For example, if I remove "layout" matrix then the controller will stop working with "default" keymap. 

Also, other controllers like kint41 or stapelberg and others implement both matrices.

Can you please explain your reasoning for removing "layout" matrix or did I misunderstand your intention?

Comment (fauxpark) [reply keyboards/kinesis/kintwin/info.json:36]: See this PR: https://github.com/qmk/qmk_firmware/pull/21384

Comment (alvicstep) [reply keyboards/kinesis/kintwin/info.json:36]: I would like to understand the request better. 

You would like to:
1. remove LAYOUT
2. create LAYOUT alias to point to LAYOUT_pretty

       Something like that:
          "layout_aliases": {
              "LAYOUT": "LAYOUT_pretty"
          },

Is it correct?

I looked at the PR you referenced. "Split" layout (or layout_pretty in our case) is easier to read and as you put it, it has "more conventionally correct ordering". And I think it is very reasonable rule to follow for new keyboards which have no existing keymaps from other contributors. And if you insist I will definitely remove "stack" layout from kintwin controller configuration. However, as I already mentioned, this will make this controller incompatible with a lot of keymaps. 

In fact, I went through every keymap under "keymaps" directory and found that there are 

- 10 keymaps which use "stack" layout: default, dvorak, dvorak_nguyenvietyen, farmergreg, heatxsink, jwon, peott-fr, stapelberg, tw1t611, xyverz

- only 3 keymaps which use "split" layout: default_pretty, insertsnideremarks, tuesdayjohn

Other contributors did a lot of work by creating these keymaps and I would like to take advantage of this work, if possible. Also, this puts kintwin controller at a disadvantage because every other controller does define both "layout" and "layout_pretty" layouts.

Comment (fauxpark) [reply keyboards/kinesis/kintwin/info.json:36]: No, `LAYOUT` should be removed, and `LAYOUT_pretty` renamed to just `LAYOUT`.
And yes, that means this PR would be blocked until those keymaps can be updated.

**Comment by alvicstep** [comment]
It is very unlikely that 10 different contributors are going to convert their keymaps to a "split" layout. They are not even aware of this effort.

Even if I can fix these 10 keymaps, I can only test them on kintwin controller. There is no guarantee they will work with the original controllers. Also, some of these keymaps are very elaborate and implement multiple layers. Converting them to a "split" design will be a very time consuming operation.

Options:
1. I can move out kintwin controller from Kinesis directory, so it will be treated as a separate keyboard. This is the approach taken by the contributor who designed "kin80" controller. This will reduce the discoverability of Kinesis based controllers and eliminate the possibility of sharing key maps among all the controllers. But it seems to be a reasonable path to move forward under these new guidelines
2. ?

File: keyboards/kinesis/kintwin/info.json
Comment (laffed) [reply keyboards/kinesis/kintwin/info.json:36]: so @alvicstep would have to edit all 10 other keymaps created by others (or they fix it on other PR's) before this PR is considered? that seems pretty unreasonable.
how is it that those 10 other keymaps were merged in to begin with. 
if a mistake was merged in, is it then the responsibility of the next PR to address those mistakes?

if a new guideline/rule was instantiated at a later date, isn't it more reasonable to open issues tagging the related parties?

Comment (fauxpark) [reply keyboards/kinesis/kintwin/info.json:36]: Have a look at my PR again.

Comment (laffed) [reply keyboards/kinesis/kintwin/info.json:36]: @fauxpark  it's unclear without digging whether your pr makes updates to keymaps that you yourself created.
what is your response to the question above?
by your last comment do you mean to say something like "i made updates to other peoples work, so you can too"?

your PR also only makes changes to existing keymaps (again unclear whether exclusively yours, exclusively not yours, or mix). which seems  more like a code improvement PR /refactor.

the scope of this PR is a new feature addition. 

again the question is: should it be that new PR's for new features have to fix previously introduced errors before they can be merged? that seems both way out of scope for one PR, and a pretty big issue for open PR congestion

Comment (fauxpark) [reply keyboards/kinesis/kintwin/info.json:36]: Yes, it is out of scope. I will sort out the user keymaps.

**Comment by alvicstep** [comment]
> It is very unlikely that 10 different contributors are going to convert their keymaps to a "split" layout. They are not even aware of this effort.
> 
> Even if I can fix these 10 keymaps, I can only test them on kintwin controller. There is no guarantee they will work with the original controllers. Also, some of these keymaps are very elaborate and implement multiple layers. Converting them to a "split" design will be a very time consuming operation.
> 
> Options:
> 
> 1. I can move out kintwin controller from Kinesis directory, so it will be treated as a separate keyboard. This is the approach taken by the contributor who designed "kin80" controller. This will reduce the discoverability of Kinesis based controllers and eliminate the possibility of sharing key maps among all the controllers. But it seems to be a reasonable path to move forward under these new guidelines
> 2. ?

File: keyboards/kinesis/kintwin/info.json
Comment (laffed) [reply keyboards/kinesis/kintwin/info.json]: @fauxpark looks like @alvicstep made changes making this conversation obsolete .
can we resolve this conversation and re-review? or are you still in the process of making commits to this branch or separate branch with the above mentioned changes?

Comment (fauxpark) [reply keyboards/kinesis/kintwin/info.json]: The keymaps have been updated. #21569

Comment (laffed) [reply keyboards/kinesis/kintwin/info.json]: @alvicstep this will nullify your last commit and prolly create a few conflicts once you rebase / merge upstream?

Comment (alvicstep) [reply keyboards/kinesis/kintwin/info.json]: Based on the changes in [21569](https://github.com/qmk/qmk_firmware/pull/21569) by @fauxpark I have moved kintwin directory back to kinesis directory.

Because 21569 PR was created off of "develop" branch I created new branch from "develop" branch as well and created new PR: [21614](https://github.com/qmk/qmk_firmware/pull/21614). 

@fauxpark , please review

P.S.
I am closing this PR. Hopefully new PR will be more lucky

**Comment by heatxsink** [comment]
What has happened?

**Pull request closed without merging**
